### ISSUE TITLE
Intermittent failure of twisted.web.test.test_webclient.WebClientSSLTestCase.test_timeoutTriggering

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3796]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#3796|
|Type|defect|
|Created|2009-04-24 15:43:47Z|
|Branch|https://github.com/twisted/twisted/tree/oldwebclient-cleanup-3796|

On OS X, I saw this fail (while working on [#3782](https://github.com/twisted/twisted/issues/3782)):

```
[ERROR]: twisted.web.test.test_webclient.WebClientSSLTestCase.test_timeoutTriggering
Traceback (most recent call last):
Failure: twisted.trial.util.DirtyReactorAggregateError: Reactor was unclean.
Selectables:
<<class 'twisted.internet.tcp.TLSConnection'> to ('127.0.0.1', 59660) at 1345af0>
```

<details><summary>Searchable metadata</summary>

```
trac-id__3796 3796
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_oldwebclient_cleanup_3796 branches/oldwebclient-cleanup-3796
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__tests tests
time__1240587827000000 1240587827000000
changetime__1310580781000000 1310580781000000
version__None None
owner__exarkun exarkun

```
</details>

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [32326]) Branching to 'oldwebclient-cleanup-3796'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

I changed `getPage` and `downloadPage` to only fire their result Deferred after the underlying connection gives the `HTTPClient` its `connectionLost` notification.

I tested this change in the [#5118](https://github.com/twisted/twisted/issues/5118) branch (with the icky `tearDown` added in that branch deleted) on my Ubuntu Karmic box to verify it actually solves the problems encountered there, and it does.  As is the branch fails 5 webclient ssl tests there, with this patch it passes them all reliably.

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/oldwebclient-cleanup-3796)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @exarkun|
|-|-|

Aah, what a nice, self-contained change.  Looks great.  Should make naive third-party tests using `getPage` a bit more reliable, too.  (And thanks for the pyflakes fixes.)

No further comments; please land.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set status to `closed`|
|-|-|

(In [32329]) Merge oldwebclient-cleanup-3796

Author: exarkun
Reviewer: glyph
Fixes: [#3796](https://github.com/twisted/twisted/issues/3796)

Change `twisted.web.client.getPage` and `twisted.web.client.downloadPage` to not
fire their returned `Deferred` until the connection used for the request each
issues is closed.  This avoids problems in unit tests using these APIs where
trial reports an unclean reactor due to these connections still being up when
the test thinks it is finished.
